plugins/ndpi/ndpi.c

```diff
@@ -0,0 +1,537 @@
+                DETECT_SM_LIST_MATCH) == NULL) {
+        goto error;
+    }
+    return 0;
+
+error:
+    if (data != NULL)
+        SCFree(data);
+    return -1;
+}
+
+static void DetectnDPIProtocolFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCFree(ptr);
+}
+
+static int DetectnDPIRiskPacketMatch(
+        DetectEngineThreadCtx *det_ctx, Packet *p, const Signature *s, const SigMatchCtx *ctx)
+{
+    const Flow *f = p->flow;
+    struct NdpiFlowContext *flowctx = FlowGetStorageById(f, flow_storage_id);
+    const DetectnDPIRiskData *data = (const DetectnDPIRiskData *)ctx;
+
+    SCEnter();
+
+    if (!flowctx->detection_completed) {
+        SCLogDebug("packet %" PRIu64 ": ndpi risks not yet detected", p->pcap_cnt);
+        SCReturnInt(0);
+    }
+
+    if (f == NULL) {
+        SCLogDebug("packet %" PRIu64 ": no flow", p->pcap_cnt);
+        SCReturnInt(0);
+    }
+
+    bool r = ((flowctx->ndpi_flow->risk & data->risk_mask) == data->risk_mask);
+    r = r ^ data->negated;
+
+    if (r) {
+        SCLogDebug("ndpi risks match on risk bitmap =  %" PRIu64 " (matching bitmap %" PRIu64 ")",
+                flowctx->ndpi_flow->risk, data->risk_mask);
+        SCReturnInt(1);
+    }
+
+    SCReturnInt(0);
+}
+
+static DetectnDPIRiskData *DetectnDPIRiskParse(const char *arg, bool negate)
+{
+    DetectnDPIRiskData *data;
+    struct ndpi_detection_module_struct *ndpi_struct;
+    ndpi_risk risk_mask;
+    NDPI_PROTOCOL_BITMASK all;
+
+    /* convert list of risk names (string) to mask */
+    ndpi_struct = ndpi_init_detection_module(NULL);
+    if (unlikely(ndpi_struct == NULL))
+        return NULL;
+
+    ndpi_struct = ndpi_init_detection_module(NULL);
+    NDPI_BITMASK_SET_ALL(all);
+    ndpi_set_protocol_detection_bitmask2(ndpi_struct, &all);
+    ndpi_finalize_initialization(ndpi_struct);
+
+    if (isdigit(arg[0]))
+        risk_mask = atoll(arg);
+    else {
+        char *dup = SCStrdup(arg), *tmp, *token;
+
+        NDPI_ZERO_BIT(risk_mask);
+
+        if (dup != NULL) {
+            token = strtok_r(dup, ",", &tmp);
+
+            while (token != NULL) {
+                ndpi_risk_enum risk_id = ndpi_code2risk(token);
+                if (risk_id >= NDPI_MAX_RISK) {
+                    SCLogError("unrecognized risk '%s', "
+                               "please check ndpiReader -H for valid risk codes",
+                            token);
+                    return NULL;
+                }
+                NDPI_SET_BIT(risk_mask, risk_id);
+                token = strtok_r(NULL, ",", &tmp);
+            }
+
+            SCFree(dup);
+        }
+    }
+
+    data = SCMalloc(sizeof(DetectnDPIRiskData));
+    if (unlikely(data == NULL))
+        return NULL;
+
+    data->risk_mask = risk_mask;
+    data->negated = negate;
+
+    return data;
+}
+
+static bool nDPIRiskDataHasConflicts(const DetectnDPIRiskData *us, const DetectnDPIRiskData *them)
+{
+    /* check for duplicate */
+    if (us->risk_mask == them->risk_mask)
```

Comment:
Suggest using `return us->risk_mask == them->risk_max;` for clarity and also making this an inline function.